File: go/transitioneventforwarding/machine.go
package main

import (
	"strconv"

	"github.com/frame-lang/frame-demos/transitioneventforwarding/framelang"
)

func NewTransitionEventForwarding(cycles int) TransitionEventForwarding {
	m := &transitionEventForwardingStruct{}

	// Validate interfaces
	var _ TransitionEventForwarding = m
	var _ TransitionEventForwarding_actions = m
	m._compartment_ = NewTransitionEventForwardingCompartment(TransitionEventForwardingState_Start)

	// Initialize domain

	// Send system start event
	params := make(map[string]interface{})
	params["cycles"] = cycles
	e := framelang.FrameEvent{Msg: ">", Params: params}
	m._mux_(&e)
	return m
}

type TransitionEventForwardingState uint

const (
	TransitionEventForwardingState_Start TransitionEventForwardingState = iota
	TransitionEventForwardingState_ForwardEventAgain
	TransitionEventForwardingState_Decrement
	TransitionEventForwardingState_Stop
)

type TransitionEventForwarding interface {
}

type TransitionEventForwarding_actions interface {
	print(msg string)
}

type transitionEventForwardingStruct struct {
	_compartment_     *TransitionEventForwardingCompartment
	_nextCompartment_ *TransitionEventForwardingCompartment
}

//====================== Multiplexer ====================//

func (m *transitionEventForwardingStruct) _mux_(e *framelang.FrameEvent) {
	switch m._compartment_.State {
	case TransitionEventForwardingState_Start:
		m._TransitionEventForwardingState_Start_(e)
	case TransitionEventForwardingState_ForwardEventAgain:
		m._TransitionEventForwardingState_ForwardEventAgain_(e)
	case TransitionEventForwardingState_Decrement:
		m._TransitionEventForwardingState_Decrement_(e)
	case TransitionEventForwardingState_Stop:
		m._TransitionEventForwardingState_Stop_(e)
	}

	if m._nextCompartment_ != nil {
		nextCompartment := m._nextCompartment_
		m._nextCompartment_ = nil
		if nextCompartment._forwardEvent_ != nil &&
			nextCompartment._forwardEvent_.Msg == ">" {
			m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
			m._compartment_ = nextCompartment
			m._mux_(nextCompartment._forwardEvent_)
		} else {
			m._do_transition_(nextCompartment)
			if nextCompartment._forwardEvent_ != nil {
				m._mux_(nextCompartment._forwardEvent_)
			}
		}
		nextCompartment._forwardEvent_ = nil
	}
}

//===================== Machine Block ===================//

func (m *transitionEventForwardingStruct) _TransitionEventForwardingState_Start_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		if e.Params["cycles"].(int) == 0 {
			m._compartment_.ExitArgs["msg"] = "stopping"
			compartment := NewTransitionEventForwardingCompartment(TransitionEventForwardingState_Stop)
			compartment._forwardEvent_ = e
			m._transition_(compartment)
			return
		} else {
			m._compartment_.ExitArgs["msg"] = "keep going"
			compartment := NewTransitionEventForwardingCompartment(TransitionEventForwardingState_ForwardEventAgain)
			compartment._forwardEvent_ = e
			m._transition_(compartment)
		}
		return
	case "<":
		m.print(e.Params["msg"].(string))
		return
	}
}

func (m *transitionEventForwardingStruct) _TransitionEventForwardingState_ForwardEventAgain_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		compartment := NewTransitionEventForwardingCompartment(TransitionEventForwardingState_Decrement)
		compartment._forwardEvent_ = e
		m._transition_(compartment)
		return
	}
}

func (m *transitionEventForwardingStruct) _TransitionEventForwardingState_Decrement_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m.print(strconv.Itoa(e.Params["cycles"].(int)))
		compartment := NewTransitionEventForwardingCompartment(TransitionEventForwardingState_Start)
		compartment.EnterArgs["cycles"] = (e.Params["cycles"].(int) - 1)
		m._transition_(compartment)
		return
	}
}

func (m *transitionEventForwardingStruct) _TransitionEventForwardingState_Stop_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m.print(strconv.Itoa(e.Params["cycles"].(int)))
		m.print("done")
		return
	}
}

//=============== Machinery and Mechanisms ==============//

func (m *transitionEventForwardingStruct) _transition_(compartment *TransitionEventForwardingCompartment) {
	m._nextCompartment_ = compartment
}

func (m *transitionEventForwardingStruct) _do_transition_(nextCompartment *TransitionEventForwardingCompartment) {
	m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
	m._compartment_ = nextCompartment
	m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

//===================== Actions Block ===================//

/********************************************************

// Unimplemented Actions

func (m *transitionEventForwardingStruct) print(msg string)  {}

********************************************************/

//=============== Compartment ==============//

type TransitionEventForwardingCompartment struct {
	State          TransitionEventForwardingState
	StateArgs      map[string]interface{}
	StateVars      map[string]interface{}
	EnterArgs      map[string]interface{}
	ExitArgs       map[string]interface{}
	_forwardEvent_ *framelang.FrameEvent
}

func NewTransitionEventForwardingCompartment(state TransitionEventForwardingState) *TransitionEventForwardingCompartment {
	c := &TransitionEventForwardingCompartment{State: state}
	c.StateArgs = make(map[string]interface{})
	c.StateVars = make(map[string]interface{})
	c.EnterArgs = make(map[string]interface{})
	c.ExitArgs = make(map[string]interface{})
	return c
}

File: csharp/TransitionEventForwarding/machine.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

using TransitionEventForwarding;
#nullable disable
namespace csharp.transitioneventforwarding
{

class TransitionEventForwarding
{

    private TransitionEventForwardingCompartment _compartment_;
    private TransitionEventForwardingCompartment _nextCompartment_;


    public TransitionEventForwarding(int cycles)
    {


        // Create and intialize start state compartment.


        this._state_ = (int)TransitionEventForwardingState.START;
        this._compartment_ = new TransitionEventForwardingCompartment(this._state_);
        this._nextCompartment_ = null;
        this._compartment_.EnterArgs["cycles"] = cycles;


        // Send system start event
        FrameEvent frameEvent = new FrameEvent(">", this._compartment_.EnterArgs);
        this._mux_(frameEvent);

    }

    // states enum
    private enum TransitionEventForwardingState
    {

        START = 0,
        FORWARDEVENTAGAIN = 1,
        DECREMENT = 2,
        STOP = 3
    }
    //====================== Multiplexer ====================//

    private void _mux_(FrameEvent e)
    {
        switch (this._compartment_.state)
        {
            case (int)TransitionEventForwardingState.START:
                this._sStart_(e);
                break;
            case (int)TransitionEventForwardingState.FORWARDEVENTAGAIN:
                this._sForwardEventAgain_(e);
                break;
            case (int)TransitionEventForwardingState.DECREMENT:
                this._sDecrement_(e);
                break;
            case (int)TransitionEventForwardingState.STOP:
                this._sStop_(e);
                break;
        }

        if( this._nextCompartment_ != null)
        {
            var nextCompartment = this._nextCompartment_;
            this._nextCompartment_ = null;
            if (nextCompartment._forwardEvent != null &&
               nextCompartment._forwardEvent._message == ">")
            {
                this._mux_(new FrameEvent( "<", this._compartment_.ExitArgs));
                this._compartment_ = nextCompartment;
                this._mux_(nextCompartment._forwardEvent);
            }
            else
            {
                this._doTransition_(nextCompartment);
                if (nextCompartment._forwardEvent != null)
                {
                    this._mux_(nextCompartment._forwardEvent);
                }
            }
            nextCompartment._forwardEvent = null;
        }
    }


    //===================== Machine Block ===================//

    TransitionEventForwardingCompartment compartment;


    private void _sStart_(FrameEvent e)
    {
        if (e._message == ">")
        {
            if (((int)e._parameters["cycles"]) == 0) {

                this._compartment_.ExitArgs["msg"] = "stopping";

                compartment =  new TransitionEventForwardingCompartment((int)TransitionEventForwardingState.STOP);

                compartment._forwardEvent = e;

                this._transition_(compartment);
                return;
            } else {

                this._compartment_.ExitArgs["msg"] = "keep going";

                compartment =  new TransitionEventForwardingCompartment((int)TransitionEventForwardingState.FORWARDEVENTAGAIN);

                compartment._forwardEvent = e;

                this._transition_(compartment);
            }
            return;
        }
        else if (e._message == "<")
        {
            this.print_do(((string)e._parameters["msg"]));
            return;
        }
    }

    private void _sForwardEventAgain_(FrameEvent e)
    {
        if (e._message == ">")
        {


            compartment =  new TransitionEventForwardingCompartment((int)TransitionEventForwardingState.DECREMENT);

            compartment._forwardEvent = e;

            this._transition_(compartment);
            return;
        }
    }

    private void _sDecrement_(FrameEvent e)
    {
        if (e._message == ">")
        {
            this.print_do(((int)e._parameters["cycles"]).ToString());


            compartment =  new TransitionEventForwardingCompartment((int)TransitionEventForwardingState.START);

            compartment.EnterArgs["cycles"]  = ((int)e._parameters["cycles"] - 1);

            this._transition_(compartment);
            return;
        }
    }

    private void _sStop_(FrameEvent e)
    {
        if (e._message == ">")
        {
            this.print_do(((int)e._parameters["cycles"]).ToString());
            this.print_do("done");
            return;
        }
    }

    //===================== Actions Block ===================//

    // Unimplemented Actions

    protected void print_do(string msg) {  
        Console.WriteLine(msg);
      }


    //=============== Machinery and Mechanisms ==============//

    private int _state_;

    private void _transition_(TransitionEventForwardingCompartment compartment)
    {
        this._nextCompartment_ = compartment;
    }

    private void _doTransition_(TransitionEventForwardingCompartment nextCompartment)
    {
        this._mux_(new FrameEvent("<", this._compartment_.ExitArgs));
        this._compartment_ = nextCompartment;
        this._mux_(new FrameEvent(">", this._compartment_.EnterArgs));
    }

}

//=============== Compartment ==============//

class TransitionEventForwardingCompartment
{

    public int state;

    public TransitionEventForwardingCompartment(int state)
    {
        this.state = state;
    }

    public Dictionary<string, object> StateArgs { get; set; } = new Dictionary<string, object>();
    public Dictionary<string, object> StateVars { get; set; } = new Dictionary<string, object>();
    public Dictionary<string, object> EnterArgs { get; set; } = new Dictionary<string, object>();
    public Dictionary<string, object> ExitArgs { get; set; } = new Dictionary<string, object>();
    public FrameEvent _forwardEvent = new FrameEvent();
}


class TransitionEventForwardingController : TransitionEventForwarding
{
        public TransitionEventForwardingController(int cycles) : base(cycles)
        {
        }

    
}

}